R/bundlePackage.R

```diff
@@ -103,73 +107,87 @@ findPackageRepoAndSource <- function(records, repos) {
+    standardizePackageSource(records[i, ], availablePackages, repos = repos)
+  })
+  rows <- lapply(rows, as.data.frame, stringsAsFactors = FALSE)
+  rbind_fill(rows, c("Source", "Repository"))
+}
+
+standardizePackageSource <- function(record, availablePackages, repos = character()) {
+  pkg <- record$Package
+  source <- record$Source
+
+  # source types are defined by packrat:
   # https://github.com/rstudio/packrat/blob/v0.9.0/R/pkg.R#L328
-  # if the package is in a named CRAN-like repository capture it
-  tmp <- lapply(seq_len(nrow(records)), function(i) {
-
-    pkg <- records[i, "Package"]
-    source <- records[i, "Source"]
-    repository <- NA
-    # capture Bioconcutor repository
-    if (identical(source, "Bioconductor")) {
-      if (pkg %in% biocPackages) {
-        repository <- biocPackages[pkg, "Repository"]
-      }
-    } else if (isSCMSource(source)) {
-      # leave source+SCM packages alone.
-    } else if (pkg %in% rownames(repo.packages)) {
-      # capture CRAN-like repository
-
-      # Find this package in the set of available packages then use its
-      # contrib.url to map back to the configured repositories.
-      package.contrib <- repo.packages[pkg, "Repository"]
-      package.repo.index <- vapply(repo.lookup$contrib.url,
-                                   function(url) grepl(url, package.contrib, fixed = TRUE), logical(1))
-      package.repo <- repo.lookup[package.repo.index, ][1, ]
-      # If the incoming package comes from CRAN, keep the CRAN name in place
-      # even if that means using a different name than the repos list.
-      #
-      # The "cran" source is a well-known location for shinyapps.io.
-      #
-      # shinyapps.io isn't going to use the manifest-provided CRAN URL,
-      # but other consumers (Connect) will.
-      if (tolower(source) != "cran") {
-        source <- package.repo$name
-      }
-      repository <- package.repo$url
+  if (source %in% c("github", "gitlab", "bitbucket")) {
+    # SCM information is recorded elsewhere
+    repository <- NA_character_
+  } else if (source == "source") {
+    # can't install source packages elsewhere
+    repository <- NA_character_
+    source <- NA_character_
+  } else if (source == "CustomCRANLikeRepository") {
+    # Package was installed from source, but packrat guessed it was installed
+    # from a known repo. We need to check that it's not actually a dev version
+    if (isDevVersion(record, availablePackages)) {
+      repository <- NA_character_
+      source <- NA_character_
+    } else {
+      repository <- findRepoUrl(pkg, availablePackages)
+      source <- findRepoName(repository, repos)
     }
-    # validatePackageSource will emit a warning for packages with NA repository.
-    data.frame(Source = source, Repository = repository, stringsAsFactors = FALSE)
-  })
-  do.call("rbind", tmp)
+  } else if (source %in% c("CRAN", "Bioconductor")) {
+    # Not strictly necessary since server will typically ignore in favour
```

Comment:
What is this comment trying to convey?

For shinyapps.io / posit.cloud, "CRAN" and "Bioconductor" are well-known sources and use server-configured URLs. For Connect, this is not the case - the URL in the manifest is used during restore.

Reply:
Oooh, I had assumed that https://docs.posit.co/connect/admin/appendix/configuration/#RPackageRepository would be set by default

Reply:
No; by default, the incoming repositories are used without modification. An admin may choose to override that choice (directing CRAN URLs to an internal RSPM mirror, for example).